File: java-advent-of-code/src/main/java/advent/twenty_fifteen/Day21RpgSimulator20xx.java
package advent.twenty_fifteen;

import advent.common.DailyProblem;
import advent.utilities.FileUtilities;
import advent.utilities.Parsers;
import com.google.common.collect.Sets;

import java.util.*;

public class Day21RpgSimulator20xx implements DailyProblem<Integer, Integer> {
    private final int part1Answer;
    private final int part2Answer;

    static class Fighter {
        int hitpoints;
        int damage;
        int armor;

        public Fighter(int hitpoints, int damage, int armor) {
            this.hitpoints = hitpoints;
            this.damage = damage;
            this.armor = armor;
        }
    }

    static class Equipment {
        String name;
        int cost;
        int damage;
        int armor;

        public Equipment(String name, int cost, int damage, int armor) {
            this.name = name;
            this.cost = cost;
            this.damage = damage;
            this.armor = armor;
        }

        @Override
        public String toString() {
            return name;
        }
    }

    private static final Set<Equipment> WEAPONS = new HashSet<>(Arrays.asList( //
            new Equipment("Dagger", 8, 4, 0), //
            new Equipment("Shortsword", 10, 5, 0), //
            new Equipment("Warhammer", 25, 6, 0), //
            new Equipment("Longsword", 40, 7, 0), //
            new Equipment("Greataxe", 74, 8, 0)));

    private static final Set<Equipment> ARMORS = new HashSet<>(Arrays.asList( //
            new Equipment("Empty Armor", 0, 0, 0), //
            new Equipment("Leather", 13, 0, 1), //
            new Equipment("Chainmail", 31, 0, 2), //
            new Equipment("Splintmail", 53, 0, 3), //
            new Equipment("Bandedmail", 75, 0, 4), //
            new Equipment("Platemail", 102, 0, 5)));

    private static final Set<Equipment> LEFT_HAND_RINGS = new HashSet<>(Arrays.asList( //
            new Equipment("Empty Left Hand", 0, 0, 0), //
            new Equipment("Damage +1", 25, 1, 0), //
            new Equipment("Damage +2", 50, 2, 0), //
            new Equipment("Damage +3", 100, 3, 0), //
            new Equipment("Defense +1", 20, 0, 1), //
            new Equipment("Defense +2", 40, 0, 2), //
            new Equipment("Defense +3", 80, 0, 3)));

    private static final Set<Equipment> RIGHT_HAND_RINGS = new HashSet<>(Arrays.asList( //
            new Equipment("Empty Right Hand", 0, 0, 0), //
            new Equipment("Damage +1", 25, 1, 0), //
            new Equipment("Damage +2", 50, 2, 0), //
            new Equipment("Damage +3", 100, 3, 0), //
            new Equipment("Defense +1", 20, 0, 1), //
            new Equipment("Defense +2", 40, 0, 2), //
            new Equipment("Defense +3", 80, 0, 3)));

    public Day21RpgSimulator20xx(String filename) {
        List<String> lines = FileUtilities.readLines(filename, Parsers::TO_STRING);

        int bossHitpoints = 0;
        int bossDamage = 0;
        int bossArmor = 0;
        for (String line : lines) {
            if (line.startsWith("Hit Points:")) {
                bossHitpoints = Integer.parseInt(line.substring(11).trim());
            }
            if (line.startsWith("Damage:")) {
                bossDamage = Integer.parseInt(line.substring(7).trim());
            }
            if (line.startsWith("Armor:")) {
                bossArmor = Integer.parseInt(line.substring(6).trim());
            }
        }

        Map<List<Equipment>, Integer> winningEquipment = new HashMap<>();
        Map<List<Equipment>, Integer> loosingEquipment = new HashMap<>();
        Set<List<Equipment>> equipmentSets = Sets.cartesianProduct(Arrays.asList(WEAPONS, ARMORS, LEFT_HAND_RINGS, RIGHT_HAND_RINGS));

        for (List<Equipment> equipment : equipmentSets) {
            Fighter me = new Fighter(100, 0, 0);

            Set<String> names = new HashSet<>();
            int cost = 0;
            for (Equipment e : equipment) {
                me.armor += e.armor;
                me.damage += e.damage;
                cost += e.cost;
                names.add(e.name);
            }
            if (names.size() != 4) {
                // Detect duplicate rings
                continue;
            }

            Fighter boss = new Fighter(bossHitpoints, bossDamage, bossArmor);

            Fighter winner = executeFight(me, boss);

            if (winner == me) {
                winningEquipment.put(equipment, cost);
            } else {
                loosingEquipment.put(equipment, cost);
            }
        }

        this.part1Answer = winningEquipment.values().stream().min(Integer::compareTo).get();
        this.part2Answer = loosingEquipment.values().stream().max(Integer::compareTo).get();
    }

    private static Fighter executeFight(Fighter lhs, Fighter rhs) {
        Fighter attacker = lhs;
        Fighter defender = rhs;
        while (true) {
            int damageDealt = Math.max(1, attacker.damage - defender.armor);
            defender.hitpoints -= damageDealt;

            if (defender.hitpoints <= 0) {
                return attacker;
            }

            Fighter temp = attacker;
            attacker = defender;
            defender = temp;
        }
    }

    @Override
    public Integer getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Integer getPart2Answer() {
        return part2Answer;
    }
}

File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day21RpgSimulator20xx.scala
package advent.twenty_fifteen

import advent.common.DailyProblem

import scala.annotation.tailrec

case class Equipment(name: String, cost: Int, damage: Int, armor: Int)

case class Fighter(name: String, hp: Int, damage: Int, armor: Int)

case class EquipmentSet(cost: Int, damage: Int, armor: Int)

case class ResultCost(win : Int, lose : Int)

class Day21RpgSimulator20xx(filename: String) extends DailyProblem[Int, Int] {

  private val WEAPONS = Set( //
    Equipment("Dagger", 8, 4, 0), //
    Equipment("Shortsword", 10, 5, 0), //
    Equipment("Warhammer", 25, 6, 0), //
    Equipment("Longsword", 40, 7, 0), //
    Equipment("Greataxe", 74, 8, 0) //
  )

  private val ARMORS = Set( //
    Equipment("Empty Armor", 0, 0, 0), //
    Equipment("Leather", 13, 0, 1), //
    Equipment("Chainmail", 31, 0, 2), //
    Equipment("Splintmail", 53, 0, 3), //
    Equipment("Bandedmail", 75, 0, 4), //
    Equipment("Platemail", 102, 0, 5) //
  )

  private val LEFT_HAND_RINGS = Set( //
    Equipment("Empty Left Hand", 0, 0, 0), //
    Equipment("Damage +1", 25, 1, 0), //
    Equipment("Damage +2", 50, 2, 0), //
    Equipment("Damage +3", 100, 3, 0), //
    Equipment("Defense +1", 20, 0, 1), //
    Equipment("Defense +2", 40, 0, 2), //
    Equipment("Defense +3", 80, 0, 3) //
  )

  private val RIGHT_HAND_RINGS = Set( //
    Equipment("Empty Right Hand", 0, 0, 0), //
    Equipment("Damage +1", 25, 1, 0), //
    Equipment("Damage +2", 50, 2, 0), //
    Equipment("Damage +3", 100, 3, 0), //
    Equipment("Defense +1", 20, 0, 1), //
    Equipment("Defense +2", 40, 0, 2), //
    Equipment("Defense +3", 80, 0, 3) //
  )

  private def combine[A](xs: Traversable[Traversable[A]]): Seq[Seq[A]] = xs.foldLeft(Seq(Seq.empty[A])) {
    (x, y) => for (a <- x; b <- y) yield a :+ b
  }

  private def runSimulations(): ResultCost = {
    @tailrec
    def fight(attacker: Fighter, defender: Fighter): String = {
      val damage = Math.max(1, attacker.damage - defender.armor)
      if (damage >= defender.hp)
        return attacker.name
      fight(Fighter(defender.name, defender.hp - damage, defender.damage, defender.armor), attacker)
    }

    val boss = Fighter("Boss", 100, 8, 2)

    combine(Set(WEAPONS, ARMORS, LEFT_HAND_RINGS, RIGHT_HAND_RINGS)) //
      .map(equip => equip.foldLeft(EquipmentSet(0, 0, 0))((acc, e) => EquipmentSet(acc.cost + e.cost, acc.damage + e.damage, acc.armor + e.armor))) //
      .foldLeft(ResultCost(Integer.MAX_VALUE, Integer.MIN_VALUE))((acc, e) => {
        if (fight(Fighter("Player", 100, e.damage, e.armor), boss).equals("Player")) {
          // Player wins
          ResultCost(Math.min(acc.win, e.cost), acc.lose)
        } else {
          // Boss wins
          ResultCost(acc.win, Math.max(acc.lose, e.cost))
        }
      }
      )
  }

  override val part1Answer: Int = runSimulations().win
  override val part2Answer: Int = runSimulations().lose
}
